DDEV project config

michalkleiner opened the pull request:
### Description

Provides DDEV project base and resolves https://github.com/matomo-org/matomo/issues/21883
Replacing community PR https://github.com/matomo-org/matomo/pull/21884 so that it's easier for us to iterate on the work, rebase, test etc.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

rr-it committed:
DDEV project base for Matomo

michalkleiner commented:
As per https://github.com/matomo-org/matomo/pull/21884#issuecomment-1927536084 for now the config files are not configured automatically:

<details>
<summary>config/config.ini.php (click to open)</summary>

```bash
; <?php exit; ?> DO NOT REMOVE THIS LINE
; file automatically generated or modified by Matomo; you can manually override the default values in global.ini.php by redefining them in this file.
[database]
host = "db"
username = "db"
password = "db"
dbname = "db"
tables_prefix = "matomo_"
charset = "utf8mb4"

[database_tests]
host = "db"
username = "root"
password = "root"

[tests]
http_host = "matomo.ddev.site"
request_uri = "/"
port = 80

[Development]
enabled = 1
disable_merged_assets = 1

[General]
enable_update_communication = 0
proxy_client_headers[] = "HTTP_X_FORWARDED_FOR"
salt = "a5ad3a95b7475e23a4b6595f6f3afa03"
trusted_hosts[] = "matomo.ddev.site"

[PluginsInstalled]
PluginsInstalled[] = "Diagnostics"
PluginsInstalled[] = "Login"
PluginsInstalled[] = "CoreAdminHome"
PluginsInstalled[] = "UsersManager"
PluginsInstalled[] = "SitesManager"
PluginsInstalled[] = "Installation"
PluginsInstalled[] = "Monolog"
PluginsInstalled[] = "Intl"
PluginsInstalled[] = "CoreVue"
PluginsInstalled[] = "CorePluginsAdmin"
PluginsInstalled[] = "CoreHome"
PluginsInstalled[] = "WebsiteMeasurable"
PluginsInstalled[] = "IntranetMeasurable"
PluginsInstalled[] = "CoreVisualizations"
PluginsInstalled[] = "Proxy"
PluginsInstalled[] = "API"
PluginsInstalled[] = "Widgetize"
PluginsInstalled[] = "Transitions"
PluginsInstalled[] = "LanguagesManager"
PluginsInstalled[] = "Actions"
PluginsInstalled[] = "Dashboard"
PluginsInstalled[] = "MultiSites"
PluginsInstalled[] = "Referrers"
PluginsInstalled[] = "UserLanguage"
PluginsInstalled[] = "DevicesDetection"
PluginsInstalled[] = "Goals"
PluginsInstalled[] = "Ecommerce"
PluginsInstalled[] = "SEO"
PluginsInstalled[] = "Events"
PluginsInstalled[] = "UserCountry"
PluginsInstalled[] = "GeoIp2"
PluginsInstalled[] = "VisitsSummary"
PluginsInstalled[] = "VisitFrequency"
PluginsInstalled[] = "VisitTime"
PluginsInstalled[] = "VisitorInterest"
PluginsInstalled[] = "RssWidget"
PluginsInstalled[] = "Feedback"
PluginsInstalled[] = "TwoFactorAuth"
PluginsInstalled[] = "CoreUpdater"
PluginsInstalled[] = "CoreConsole"
PluginsInstalled[] = "ScheduledReports"
PluginsInstalled[] = "UserCountryMap"
PluginsInstalled[] = "Live"
PluginsInstalled[] = "PrivacyManager"
PluginsInstalled[] = "ImageGraph"
PluginsInstalled[] = "Annotations"
PluginsInstalled[] = "MobileMessaging"
PluginsInstalled[] = "Overlay"
PluginsInstalled[] = "SegmentEditor"
PluginsInstalled[] = "Insights"
PluginsInstalled[] = "Morpheus"
PluginsInstalled[] = "Contents"
PluginsInstalled[] = "TestRunner"
PluginsInstalled[] = "BulkTracking"
PluginsInstalled[] = "Resolution"
PluginsInstalled[] = "DevicePlugins"
PluginsInstalled[] = "Heartbeat"
PluginsInstalled[] = "Marketplace"
PluginsInstalled[] = "ProfessionalServices"
PluginsInstalled[] = "UserId"
PluginsInstalled[] = "CustomJsTracker"
PluginsInstalled[] = "Tour"
PluginsInstalled[] = "PagePerformance"
PluginsInstalled[] = "CustomDimensions"
PluginsInstalled[] = "JsTrackerInstallCheck"

```

</details>

<details>
<summary>tests/UI/config.js (click to open)</summary>

```js
/*!
 * Matomo - free/libre analytics platform
 *
 * UI tests config
 *
 * @link https://matomo.org
 * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
 */

/**
 * The root Piwik URL to test against.
 */
exports.piwikUrl = "http://matomo.ddev.site/";

/**
 * Data for the $_SERVER variable in the setup/teardown PHP scripts. Should be the same as
 * the values in your phpunit.xml file.
 */
exports.phpServer = {
    HTTP_HOST: 'matomo.ddev.site',
    REQUEST_URI: '/',
    REMOTE_ADDR: '127.0.0.1'
};

/**
 * The path to the PHP executable to execute when setting up & tearing down the database.
 */
exports.php = 'php';

/**
 * The mocha reporter to use.
 */
exports.reporter = "spec";

/**
 * The directory that stores expected screenshots. Relative to the UI repo's root directoriy.
 */
exports.expectedScreenshotsDir = ["./expected-screenshots", "./expected-ui-screenshots"];

/**
 * The directory that stores processed screenshots. Relative to the UI repo's root directory.
 */
exports.processedScreenshotsDir = "./processed-ui-screenshots";

/**
 * The directory that stores screenshot diffs. Relative to the UI repo's root directory.
 */
exports.screenshotDiffDir = "./screenshot-diffs";

```

</details>

sgiehl commented:
I haven't tested anything yet, but here are already some thoughts I had while looking at the code:
* If we want to have the ddev config part of this repo, we need to ensure those files are removed before building a release
* Assuming that the actual code is mounted into the vm: Running npm install will fail on windows in this case, as directory depth is kind of always too deep for the file system and in many cases it tries to create symlinks, which isn't supported on windows. A possible workaround is to hold the `node_module` folders fully inside the vm by e.g. overwriting the path with internal mount points or similar.

michalkleiner committed:
Ignore tests' node_modules

michalkleiner committed:
Add ll alias

michalkleiner committed:
Ensure DDEV files are not part of a release

michalkleiner committed:
Add console command wrapper

michalkleiner committed:
Run base init after env startup

michalkleiner committed:
Fix typos in main readme

michalkleiner commented:
Ok, I think this is now in a testable state with some documentation needed.
Would someone be keen to check out this branch and test if these steps work?

1. clone the repo, cd to the repo folder and checkout `dev-18030-ddev-env-config` branch
2. run `ddev start` to start the project up
3. run `ddev launch` to open the instance in browser
4. follow the installation process (db details should be pre-filled) and log in
5. run `ddev matomo:init:tests` and see if all the install steps work
6. run `ddev matomo:console tests:run` to see if tests work
7. run `ddev matomo:console tests:run-ui` to see if UI tests work

mneudert commented on .ddev/commands/host/matomo_init_tests:
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
```suggestion
#!/usr/bin/env bash
```

For all the host commands.

My NixOS box only has `/bin/sh` and `/usr/bin/env` available, no `/bin/bash`. I would not expect someone to have this problem, but `/usr/bin/env` should be just a tiny bit more portable.

We could also switch to plain `/bin/sh`, but that may require changes to the `read` arguments afaik.

mneudert commented on .ddev/config.yaml:
```diff
@@ -0,0 +1,296 @@
+    version: "5.7"
+hooks:
+    post-start:
+        - exec: MYSQL_PWD=root mysql -uroot -e "CREATE DATABASE IF NOT EXISTS matomo_tests; GRANT ALL ON matomo_tests.* to 'db'@'%';"
+          service: db
+        - exec: MYSQL_PWD=root mysql -uroot -e "CREATE DATABASE IF NOT EXISTS _UITestFixture; GRANT ALL ON _UITestFixture.* to 'db'@'%';"
+          service: db
```
```suggestion
        - exec: MYSQL_PWD=root mysql -uroot -e "GRANT ALL PRIVILEGES ON *.* to 'db'@'%'"
          service: db
        - exec: MYSQL_PWD=root mysql -uroot -e "GRANT ALL PRIVILEGES ON *.* to 'db'@'localhost'"
          service: db
```

This would allow us to configure `db:db` for the tests instead of `root:root`.

Otherwise there may be databases not visible to the `db` user, for example when running when running the `EmptySite` spec:

```shell
ddev matomo:console tests:run-ui --persist-fixture-data EmptySite
```

Without more privileges the created `_EmptySite` database would not be visible.

_Not 100% sure if we need to run the statement for if for `@localhost`, I think it was required to have both `ddev mysql` and `mysql` to have all the privileges.._

michalkleiner replied:
I wouldn't expect the `@localhost` to be needed as the db is not running on the same system so PHP will never access it on localhost.

I'll test it out, thanks for pointing this out and identifying a test that can be impacted by this.

mneudert commented on .ddev/config.yaml:
```diff
@@ -0,0 +1,296 @@
+        - exec: MYSQL_PWD=root mysql -uroot -e "CREATE DATABASE IF NOT EXISTS _UITestFixture; GRANT ALL ON _UITestFixture.* to 'db'@'%';"
+          service: db
+        - exec-host: ddev matomo:init
+webimage_extra_packages: [build-essential, chromium, fonts-dejavu, fonts-liberation, gconf-service, imagemagick, imagemagick-doc, libappindicator1, libasound2, libatk1.0-0, libcairo2, libdrm2, libgbm1, libgconf-2-4, libgdk-pixbuf2.0-0, libgtk-3-0, libnspr4, libnss3, libpango-1.0-0, libpangocairo-1.0-0, libx11-xcb1, libxcomposite1, libxcursor1, libxdamage1, libxfixes3, libxi6, libxrandr2, libxrend ...
```
Some of the screenshot tests give really weird results. PhpStorm shows all processed screenshots as "24-bit" instead of "32-bit".

And some screenshots created in the Dashboard spec (e.g. `Dashboard_dashboard1_mobile.png`) are stretched beyond recognition, though the height/width of the file match the expected screenshots.

michalkleiner replied:
24 vs 32 bit wouldn't be a problem if we eventually run CI using ddev as well, as long as the images carry the visual information they should.

Could you please send me some of the image names or even some of the files so that I could compare them with mine?

mneudert replied:
Sent you my local images and some additional information. It looks like the DDEV environment generates images without an alpha channel:

```
// repository
  Type: TrueColorAlpha
  Endianness: Undefined
  Depth: 8-bit
  Channels: 4.0
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
    Alpha: 1-bit

// ddev
  Type: TrueColor
  Endianness: Undefined
  Depth: 8-bit
  Channels: 3.0
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
```

rr-it commented:
I had some issue with starting the web-container caused by `supervisord`: `PermissionError: [Errno 13] Permission denied: '/var/log/supervisord.log'`
Also see https://github.com/ddev/ddev/issues/4899

Adding this file `.ddev/web-entrypoint.d/fix-var-log-directory-permissions.sh` solved it:

```bash
#!/bin/bash

# Make mounted `/var/log` volume writable for supervisord.
sudo chmod 777 /var/log

```

mneudert committed:
Update host command hashbangs

mneudert committed:
Ensure "local_infile=ON" is set

MySQL 5.7 has this as the default value, but MySQL changed to
"local_infile=OFF". Configuring this explicitly allows more seamless
transition between databases using "ddev debug migrate-database".

mneudert committed:
Run tests with "db:db" database credentials

mneudert committed:
Force rewrite of config during "matomo:init:tests"

Running the init script multiple times can generate lots of duplicate
sections in the file. Forcing a change triggers a rewrite of the config
file, cleaning up all duplicates, and reordering the sections.

mneudert committed:
Install "ripgrep" for "tests:check-direct-dependency-use"

mneudert committed:
Mix and match DDEV+Matomo xhprof setups

Will use the DDEV provided xhprof module, but all profiling controlled
by Matomo using "[Debug] enable_php_profiler=1", CLI parameters, and URL
parameters.

The GUI will be served using the DDEV version, with an alias on the
"/vendor/xhprof" path to keep the required changes to a minimum for now.

mneudert committed:
Add README for DDEV environment

michalkleiner commented on .ddev/xhprof/xhprof_prepend.php:
```diff
@@ -0,0 +1,3 @@
+<?php
+
+// Matomo provides it's own xhprof integration.
```
```suggestion
// Matomo provides its own xhprof integration.
```

michalkleiner commented on .ddev/config.yaml:
```diff
@@ -0,0 +1,292 @@
+name: playground
```
```suggestion
name: matomo
```

michalkleiner commented on .ddev/config.yaml:
```diff
@@ -0,0 +1,292 @@
+additional_hostnames: []
+additional_fqdns: []
+database:
+    type: mariadb
+    version: "11.4"
```
```suggestion
    type: mysql
    version: "5.7"
```

michalkleiner commented on .ddev/apache/apache-site.conf:
```diff
@@ -0,0 +1,99 @@
+# If you want to take over this file and customize it, remove the line above
+# and ddev will respect it and won't overwrite the file.
+# See https://ddev.readthedocs.io/en/stable/users/extend/customization-extendibility/#custom-apache-configuration
```
```suggestion
# This file is customised for Matomo, the differences are related to XHProf.
#
# See https://ddev.readthedocs.io/en/stable/users/extend/customization-extendibility/#custom-apache-configuration
```

michalkleiner commented on .ddev/commands/host/matomo_init:
```diff
@@ -0,0 +1,19 @@
+## Usage: matomo:init
+## Example: ddev matomo:init
+
+#echo "Copy initial config file ..."
+#echo "@see https://developer.matomo.org/guides/tests-php#requirements"
+#ddev exec '\cp -r .ddev/initial-config/config.ini.php config/config.ini.php'
```
I think it's ok to remove these, but we should call a bunch of `ddev matomo:console config:set` and set the config values from the initial ini config.

mneudert replied:
As with several other changes this was not really intended, but a result of local modifications and a really bad `git add` :see_no_evil: 

```suggestion
echo "Copy initial config file ..."
echo "@see https://developer.matomo.org/guides/tests-php#requirements"
ddev exec '\cp -r .ddev/initial-config/config.ini.php config/config.ini.php'
```

michalkleiner commented on .ddev/initial-config/config.ini.php.partial:
As mentioned in the comment for the `matomo:init` command, we might be able to replace this template by a set of `config:set` calls. That should also allow us to remove the rewriting just to normalise the config file.

michalkleiner commented on mysqldefaults.conf:
```diff
@@ -0,0 +1,3 @@
+[client]
+user=root
+password=root
```
Would this be `db:db` now?

mneudert replied:
This file should not even be added to the repository here. We could add it to `.gitignore`, but not just create one.

michalkleiner commented on package-lock.json:
Probably should reset this, the env wouldn't add any new or update existing npm dependencies.

michalkleiner commented on plugins/Goals/tests/Integration/APITest.php:
```diff
@@ -19,6 +19,7 @@
  * @group Plugins
  * @group APITest
  * @group Plugins
+ * @group RunMeOnce
```
```suggestion
```

michalkleiner commented on .ddev/README.md:
```diff
@@ -0,0 +1,56 @@
+# Matomo DDEV Environment
```
```suggestion
# Matomo DDEV environment
```

michalkleiner commented on .ddev/README.md:
```diff
@@ -0,0 +1,56 @@
+
+Before you begin, make sure you have DDEV installed. If you don't have it yet, follow the installation instructions in the official [DDEV documentation](https://ddev.readthedocs.io/en/stable/).
+
+## Setup Steps
+
+### 1. Start the Environment
```
```suggestion
## Setup steps

### 1. Start the environment
```

michalkleiner commented on .ddev/README.md:
````diff
@@ -0,0 +1,56 @@
+ddev start
+```
+
+This command will start the DDEV environment for local development. By default, your local environment will be accessible at [matomo.ddev.site](https://matomo.ddev.site).
````
```suggestion
This command will start the DDEV environment for local development. By default, your local environment will be accessible at [https://matomo.ddev.site](https://matomo.ddev.site).
```

michalkleiner commented on .ddev/README.md:
```diff
@@ -0,0 +1,56 @@
+
+This will automatically open the default browser and take you to the Matomo setup page.
+
+### 2. Set Up Matomo Using the UI Installer
```
```suggestion
### 2. Set up Matomo using the UI installer
```

michalkleiner commented on .ddev/README.md:
```diff
@@ -0,0 +1,56 @@
+
+Follow the on-screen instructions in the UI installer to complete the Matomo setup. This will configure the necessary database and settings for your local instance.
+
+### 3. Set Up the Testing Environment
```
```suggestion
### 3. Set up the testing environment
```

michalkleiner commented on .ddev/README.md:
```diff
@@ -0,0 +1,56 @@
+
+For more information about Matomo development, check out the official [Matomo Developer Documentation](https://developer.matomo.org/).
+
+## Known Issues
```
```suggestion
## Known issues
```

michalkleiner reviewed (commented):
Thanks for all the testing @mneudert! I've left a few comments to action, but we should be ready to get this in soon! Can't wait! 🚀

michalkleiner commented:
We also need to force use of Node 16 so perhaps in the main init (or the tests init) we need `ddev nvm install 16`. Probably in the main init since running `vue:build` should work even without all the initialisations. Then also we may change `ddev matomo:init:tests` to `ddev matomo:init:dev` since that also installs npm dependencies which are required for `vue:build` which is not necessarily _just_ tests, it is handy for development as well. `ddev matomo:init` should still just be the bare minimum to be able to run Matomo locally, so I wouldn't put it there to keep the footprint there minimal. Originally I had 3 init mechanisms but that would be too many I reckon.

mneudert committed:
Apply suggestions from code review

Co-authored-by: Michal Kleiner <michal@innocraft.com>

mneudert committed:
Use config:set instead of fixed config file for test init

mneudert committed:
Remove outdated config file copy

mneudert committed:
Split dev/test initialisation

mneudert commented on .ddev/commands/host/matomo_init_dev:
```diff
@@ -0,0 +1,16 @@
+
+## Description: Initialize Matomo environment for development
+## Usage: matomo:init:dev
+## Example: ddev matomo:init:dev
```
Is this the split you had in mind @michalkleiner?

The "base init" should not be necessary as it is done on every startup. And,, unless my local environment acted up, testing can run without the development configuration active.

michalkleiner replied:
Yep, tested and tests can be run without the dev init, matomo itself can run without both dev and tests inits so 👍

mneudert commented on .ddev/config.yaml:
```diff
@@ -0,0 +1,292 @@
+    - MATOMO_DATABASE_USERNAME=db
+    - MATOMO_DATABASE_PASSWORD=db
+    - MATOMO_DATABASE_DBNAME=db
+nodejs_version: "16"
```
This should set the system NodeJS version to 16.x, allowing to run `vue:build` right away after startup.

michalkleiner replied:
works as expected

michalkleiner committed:
Add warning not to use to host prod instances

michalkleiner reviewed (commented):
Great collaboration 🚀
